modules/sdk-coin-sui/src/lib/stakingBuilder.ts

```diff
@@ -0,0 +1,162 @@
+    return TransactionType.AddDelegator;
+  }
+
+  /**
+   * Create a new transaction for delegating coins ready to be signed and executed.
+   *
+   * @param addDelegationTx
+   */
+  requestAddDelegation(addDelegationTx: RequestAddDelegationTx): this {
+    this.validateAddress({ address: addDelegationTx.validatorAddress });
+    this.validateValue(BigNumber(addDelegationTx.amount));
+    for (const coin of addDelegationTx.coins) {
+      this.validateSuiObjectRef(coin, 'addDelegationTx.coins');
+    }
+    this._addDelegationTx = addDelegationTx;
+    this._moveCallTx = {
+      package: SUI_PACKAGE,
+      module: ModulesNames.SuiSystem,
+      function: MethodNames.RequestAddDelegationMulCoin,
+      typeArguments: [],
+      arguments: [
+        SUI_SYSTEM_STATE_OBJECT,
+        addDelegationTx.coins,
+        addDelegationTx.amount,
+        addDelegationTx.validatorAddress,
+        // utils.mapSharedObjectToCallArg(SUI_SYSTEM_STATE_OBJECT),
+        // utils.mapCoinsToCallArg(addDelegationTx.coins),
+        // utils.mapAmountToCallArg(addDelegationTx.amount),
+        // utils.mapAddressToCallArg(addDelegationTx.validatorAddress),
+      ],
+    };
+    return this;
+  }
+
+  /** @inheritdoc */
+  protected fromImplementation(rawTransaction: string): Transaction<MoveCallTx> {
+    const tx = new StakingTransaction(this._coinConfig);
+    this.validateRawTransaction(rawTransaction);
+    tx.fromRawTransaction(rawTransaction);
+    this.initBuilder(tx);
+    return this.transaction;
+  }
+
+  /** @inheritdoc */
+  protected async buildImplementation(): Promise<Transaction<MoveCallTx>> {
+    this.transaction.setSuiTransaction(this.buildSuiTransaction());
+    this.transaction.transactionType(this.transactionType);
+    this.transaction.loadInputsAndOutputs();
+    return this.transaction;
+  }
+
+  /**
+   * Initialize the transaction builder fields using the decoded transaction data
+   *
+   * @param {Transaction} tx the transaction data
+   */
+  initBuilder(tx: StakingTransaction): void {
+    this._transaction = tx;
+    this._signatures = [tx.suiSignature];
+
+    const txData = tx.toJson();
+    this.gasBudget(txData.gasBudget);
+    this.sender(txData.sender);
+    this.gasPayment(txData.gasPayment);
+
+    const txDetails = txData.kind.Single as MoveCallTxDetails;
+    if (txDetails.hasOwnProperty('Call')) {
+      if (txDetails.Call.function === MethodNames.RequestAddDelegationMulCoin) {
+        this.type(SuiTransactionType.AddDelegation);
+        this.requestAddDelegation({
+          coins: txDetails.Call.arguments[1] as SuiObjectRef[],
+          amount: Number(txDetails.Call.arguments[2]),
+          validatorAddress: txDetails.Call.arguments[3].toString(),
+        });
+      } else {
+        throw new NotImplementedError(`${txDetails.Call.function} not implemented`);
+      }
+    } else {
+      throw new Error('Transaction type not supported: ' + txDetails);
+    }
```

Comment:
NIT we can avoid unnecessary if else chaining 
```js
   if (!txDetails.hasOwn('Call')) {
        throw new Error('Transaction type not supported: ' + txDetails);
   }
   
   if (txDetails.Call.function !== MethodNames.RequestAddDelegationMulCoin) {
        throw new NotImplementedError(`${txDetails.Call.function} not implemented`);
   }
   
    this.type(SuiTransactionType.AddDelegation);
    this.requestAddDelegation({
          coins: txDetails.Call.arguments[1] as SuiObjectRef[],
          amount: Number(txDetails.Call.arguments[2]),
          validatorAddress: txDetails.Call.arguments[3].toString(),
      });
      
```
Also seems **hasOwn** more recommended than hasOwnProperty. see [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn)

Reply:
fixing...